controllers/storageclassclaim/storageclassclaim_controller.go

```diff
@@ -99,11 +137,481 @@ func (r *StorageClassClaimReconciler) reconcileConsumerPhases(ctx context.Contex
 }
 
 func (r *StorageClassClaimReconciler) reconcileProviderPhases(ctx context.Context, storagecluster *v1.StorageCluster) (reconcile.Result, error) {
-	log := ctrllog.FromContext(ctx)
-	log.Info("Running StorageClassClaim controller in Converged/Provider Mode")
+	if !r.storageConsumer.Spec.Enable {
+		r.storageClassClaim.Status.Phase = string(v1alpha1.StorageClassClaimStateDisabled)
+		return reconcile.Result{}, nil
+	}
+
+	r.storageClassClaim.Status.Phase = string(v1alpha1.StorageClassClaimStateConfiguring)
+
+	for _, cephResourceSpec := range r.storageClassClaim.Status.CephResources {
+		r.cephResourcesByName[cephResourceSpec.Name] = cephResourceSpec
+	}
+
+	if r.storageClassClaim.GetDeletionTimestamp().IsZero() {
+		if !contains(r.storageClassClaim.GetFinalizers(), storageClassClaimFinalizer) {
+			r.log.Info("Finalizer not found for StorageClassClaim. Adding finalizer.", "StorageClassClaim", klog.KRef(r.storageClassClaim.Namespace, r.storageClassClaim.Name))
+			r.storageClassClaim.ObjectMeta.Finalizers = append(r.storageClassClaim.ObjectMeta.Finalizers, storageClassClaimFinalizer)
+			if err := r.Client.Update(ctx, r.storageClassClaim); err != nil {
+				r.log.Error(err, "Failed to update StorageClassClaim with finalizer.", "StorageClassClaim", klog.KRef(r.storageClassClaim.Namespace, r.storageClassClaim.Name))
+				return reconcile.Result{}, err
+			}
+		}
+		if r.storageClassClaim.Spec.Type == "cephBlockPool" {
+			if err := r.reconcileCephClientRBDProvisioner(); err != nil {
+				return reconcile.Result{}, err
+			}
+
+			if err := r.reconcileCephClientRBDNode(); err != nil {
+				return reconcile.Result{}, err
+			}
+
+			if err := r.reconcileCephBlockPool(storagecluster); err != nil {
+				return reconcile.Result{}, err
+			}
+		} else if r.storageClassClaim.Spec.Type == "cephFileSystemSubVolumeGroup" {
+			if err := r.reconcileCephClientCephFSProvisioner(); err != nil {
+				return reconcile.Result{}, err
+			}
+
+			if err := r.reconcileCephClientCephFSNode(); err != nil {
+				return reconcile.Result{}, err
+			}
+
+			if err := r.reconcileCephFilesystemSubVolumeGroup(); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+		cephResourcesReady := true
+		for _, cephResource := range r.storageClassClaim.Status.CephResources {
+			if cephResource.Phase != "Ready" {
+				cephResourcesReady = false
+				break
+			}
+		}
+
+		if cephResourcesReady {
+			r.storageClassClaim.Status.Phase = string(v1alpha1.StorageClassClaimStateReady)
+		}
+
+	} else {
+		// todo deletion of the Ceph Resources
+	}
+
+	r.log.Info("Running StorageClassClaim controller in Converged/Provider Mode")
```

Comment:
This should be the first line of the fuction.

Reply:
ack